server/src/main/java/com/linecorp/centraldogma/server/internal/MetadataRolesMigrationPlugin.java

```diff
@@ -0,0 +1,127 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.centraldogma.server.internal;
+
+import java.time.Instant;
+import java.util.concurrent.CompletionStage;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.common.Change;
+import com.linecorp.centraldogma.common.Entry;
+import com.linecorp.centraldogma.common.Markup;
+import com.linecorp.centraldogma.common.Query;
+import com.linecorp.centraldogma.common.Revision;
+import com.linecorp.centraldogma.internal.Jackson;
+import com.linecorp.centraldogma.server.CentralDogmaConfig;
+import com.linecorp.centraldogma.server.command.Command;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.command.CommitResult;
+import com.linecorp.centraldogma.server.metadata.MetadataService;
+import com.linecorp.centraldogma.server.plugin.Plugin;
+import com.linecorp.centraldogma.server.plugin.PluginContext;
+import com.linecorp.centraldogma.server.plugin.PluginTarget;
+import com.linecorp.centraldogma.server.storage.project.InternalProjectInitializer;
+import com.linecorp.centraldogma.server.storage.project.Project;
+import com.linecorp.centraldogma.server.storage.repository.Repository;
+
+public final class MetadataRolesMigrationPlugin implements Plugin {
+
+    private static final Logger logger = LoggerFactory.getLogger(MetadataRolesMigrationPlugin.class);
+
+    @VisibleForTesting
+    static final String METADATA_ROLES_MIGRATION_JOB_LOG =
+            "/metadata-roles-job.json";
+
+    @Override
+    public PluginTarget target(CentralDogmaConfig config) {
+        return PluginTarget.LEADER_ONLY;
+    }
+
+    @Override
+    public CompletionStage<Void> start(PluginContext context) {
+        if (hasMigrationLog(context)) {
+            logger.debug("Metadata roles have already been migrated. Skipping auto migration...");
+            return UnmodifiableFuture.completedFuture(null);
+        }
+        logger.info("Starting metadata roles migration ...");
+
+        final MetadataService metadataService = new MetadataService(context.projectManager(),
+                                                                    context.commandExecutor());
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        for (Project project : context.projectManager().list().values()) {
+            final String name = project.name();
```

Comment:
Should we add a migration log for each project so we can identify the progress and investegate when something goes wrong?

Reply:
It's a relatively simple migration so I thought just leaving the warning log was enough. 😄